### PR TITLE
add dirichlet random sample op in cpu and gpu kernel

### PR DESCRIPTION
**Pull request by cxxly**
### PR types

New features
### PR changes

OPs
### Describe

Dirichlet and Beta distribution API need fast dirichlet random generator for draw sample data,  we implement it in cpu and cuda backend use [reject-accept](https://dl.acm.org/doi/10.1145/358407.358414) sample and[ inverse-transform](https://en.wikipedia.org/wiki/Inverse_transform_sampling) sample method.

### CODE REVIEW

File: paddle/fluid/operators/dirichlet_op.h [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+#else
+#define compat_exp std::exp
+#define compat_ceil std::ceil
+#define compat_floor std::floor
+#define compat_log std::log
+#define compat_pow std::pow
+#define compat_sqrt std::sqrt
+#define compat_tan std::tan
+#define compat_abs std::abs
+#define compat_log1p std::log1p
```
Comment (jeff41404) [inline paddle/fluid/operators/dirichlet_op.h]: macro's name use underline delimited all capital charactor nomenclature

File: paddle/fluid/operators/dirichlet_op.h [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+#endif
+
+namespace paddle {
+namespace operators {
+template <typename DeviceContext, typename T>
+struct DirichletSampler;
+
+template <typename scalar_t, typename sampler_t>
```
Comment (jeff41404) [inline paddle/fluid/operators/dirichlet_op.h]: The first letter of template parameter should be capitalized.

File: paddle/fluid/operators/dirichlet_op.cu [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#if defined(PADDLE_WITH_CUDA)
+using compatRandStatePhilox4_32_10_t = curandStatePhilox4_32_10_t;
+#define compat_rand_init curand_init
+#define compat_rand_uniform curand_uniform
+#define compat_rand_normal curand_normal
+#elif defined(PADDLE_WITH_HIP)
+using compatRandStatePhilox4_32_10_t = hiprandStatePhilox4_32_10_t;
+#define compat_rand_init hiprand_init
+#define compat_rand_uniform hiprand_uniform
+#define compat_rand_normal hiprand_normal
```
Comment (jeff41404) [inline paddle/fluid/operators/dirichlet_op.cu]: macro's name use underline delimited all capital charactor nomenclature

File: paddle/fluid/operators/dirichlet_op.h [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+template <typename scalar_t, typename accscalar_t, typename uniform_sampler_t,
+          typename normal_sampler_t>
```
Comment (jeff41404) [inline paddle/fluid/operators/dirichlet_op.h]: The first letter of template parameter should be capitalized.

File: paddle/fluid/operators/dirichlet_op.cc [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+#include "paddle/fluid/operators/dirichlet_op.h"
+
+#include "paddle/fluid/framework/generator.h"
+#include "paddle/fluid/operators/elementwise/elementwise_op_function.h"
+#include "paddle/fluid/operators/reduce_ops/reduce_op.h"
+#include "paddle/fluid/operators/reduce_ops/reduce_sum_op.h"
+
+namespace paddle {
+namespace operators {
+template <typename T, typename uniform_sampler_t, typename normal_sampler_t>
```
Comment (jeff41404) [inline paddle/fluid/operators/dirichlet_op.cc]: The first letter of template parameter should be capitalized.

File: paddle/fluid/operators/dirichlet_op.h (hunk 1 above)
Comment (cxxly) [reply paddle/fluid/operators/dirichlet_op.h]: updated

File: paddle/fluid/operators/dirichlet_op.h (hunk 2 above)
Comment (cxxly) [reply paddle/fluid/operators/dirichlet_op.h]: have updated

File: paddle/fluid/operators/dirichlet_op.cu (hunk 3 above)
Comment (cxxly) [reply paddle/fluid/operators/dirichlet_op.cu]: have updated

File: paddle/fluid/operators/dirichlet_op.h (hunk 4 above)
Comment (cxxly) [reply paddle/fluid/operators/dirichlet_op.h]: have updated

File: paddle/fluid/operators/dirichlet_op.cc (hunk 5 above)
Comment (cxxly) [reply paddle/fluid/operators/dirichlet_op.cc]: have updated

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**